Layers docs (#32)
caikit/website: docs/architecture_layers.md

---
title: Architecture Layers
layout: default
nav_order: 6
---

# Architecture Layers

The `caikit` project is itself a stack of complimentary building blocks that layer on top of each other to provide rising levels of abstraction to the various functions in the space of AI problems. The layers roughly break down as follows:

> 1. [**Useful Tools**](#1-useful-tools): At the bottom, `caikit` holds a collection of tools that are not AI-specific. These tools are useful for AI, but can be imported and used in just about any problem space where they fit.

Author: Nice, I like that clarification

    DataBuffer.from_iterable([4, 5, 6]),
    DataBuffer.from_iterable([0.7, 0.8, 0.9]),
])

print("## To Types")
print(embedding_set.to_types())
print()

print("## To Json")
print(embedding_set.to_json())
print()

print("## To Protobuf")
print(embedding_set.to_proto())
print()
```

```sh
python dataobject_example.py
# >>> ## To Types
# >>> [(1, 2, 3), (4, 5, 6), (0.7, 0.8, 0.9)]
# >>>
# >>> ## To Json
# >>> {"embeddings": [{"data": "AAAAAz/wAAAAAAAAQAAAAAAAAABACAAAAAAAAA==", "data_type": "INT"}, {"data": "AAAAA0AQAAAAAAAAQBQAAAAAAABAGAAAAAAAAA==", "data_type": "INT"}, {"data": "AAAAAz/mZmZmZmZmP+mZmZmZmZo/7MzMzMzMzQ==", "data_type": "FLOAT"}]}
# >>>
# >>> ## To Protobuf
# >>> embeddings {
# >>>   data: "\000\000\000\003?\360\000\000\000\000\000\000@\000\000\000\000\000\000\000@\010\000\000\000\000\000\000"
# >>>   data_type: INT
# >>> }
# >>> embeddings {
# >>>   data: "\000\000\000\003@\020\000\000\000\000\000\000@\024\000\000\000\000\000\000@\030\000\000\000\000\000\000"
# >>>   data_type: INT
# >>> }
# >>> embeddings {
# >>>   data: "\000\000\000\003?\346ffffff?\351\231\231\231\231\231\232?\354\314\314\314\314\314\315"
# >>>   data_type: FLOAT
# >>> }
```

### Data Streaming

> For many AI workloads, especially `training`, data volumes quickly escape the bounds of in-memory iterables. There are numerous solutions for managing iterables of data backed by disk or even remote services. In `caikit`, these are all managed via the [DataStream](https://github.com/caikit/caikit/blob/main/caikit/core/data_model/streams/data_stream.py) utility. This class provides consistent iteration semantics wrapping arbitrary [python generators](https://wiki.python.org/moin/Generators).

Author: Good catch. The former!

### Domain Data Model

Each sub module within `caikit.interfaces.<domain>` holds a `data_model` module that defines the key [dataobject](#data-modeling) structs for the tasks within the given domain.

### Domain Tasks

Each sub module within `caikit.interfaces.<domain>` holds a `tasks` module that defines the key [task signatures](#tasks) for the problems to be solved in the given domain.

## 5. AI Domain Libraries

Within the `caikit` project, there is an evolving set of libraries that offer concrete implementations of [tasks](#tasks) based on their domains.

### Caikit NLP

> The most mature AI domain in the `caikit` community is Natural Language Processing (NLP). The [caikit-nlp project](https://github.com/caikit/caikit-nlp) holds modules implementing the most essential tasks in NLP, particular for generative AI usecases.

Reviewer: ```suggestion
The most mature AI domain in the `caikit` community is Natural Language Processing (NLP). The [caikit-nlp project](https://github.com/caikit/caikit-nlp) holds modules implementing the most essential tasks in NLP, particularly for generative AI use cases.
```

* [**Text Generation**](https://github.com/caikit/caikit-nlp/tree/main/caikit_nlp/modules/text_generation): Generative models which take prompt text in and generate novel text based on the prompt.
* [**Text Embedding**](https://github.com/caikit/caikit-nlp/tree/main/caikit_nlp/modules/text_embedding): Given input prompt text, generate embedding vectors based on a language model
* [**Tokenization**](https://github.com/caikit/caikit-nlp/tree/main/caikit_nlp/modules/tokenization): Given a string of text, split the text into a sequence of individual tokens

> * [**Token Classification**](https://github.com/caikit/caikit-nlp/tree/main/caikit_nlp/modules/token_classification): Given a sequence of tokens, detect token groups which match certain classes (e.g. Personally Identifiable Information)

Reviewer: ```suggestion
* [**Token Classification**](https://github.com/caikit/caikit-nlp/tree/main/caikit_nlp/modules/token_classification): Given a sequence of tokens, detect token groups that match certain classes (e.g. Personally Identifiable Information)
```

* [**Text Classification**](https://github.com/caikit/caikit-nlp/tree/main/caikit_nlp/modules/text_classification): Given a string of text, classify it based on a known set of classes

### Caikit Computer Vision

The `caikit-computer-vision` is in its early stages. It currently holds a single `transformers` implementation of the [**Object Detection**](https://github.com/caikit/caikit-computer-vision/tree/main/caikit_computer_vision/modules/object_detection) task, but will continue to grow as usage and contributions add up.

## 6. Prebuilt Runtime Images

The `caikit` community, in close partnership with Red Hat's OpenShift AI team, provides several ways to build/consume container images that are ready to use in a runtime of the user's choosing.

### Caikit NLP

> The `caikit-nlp` project provides a [Dockerfile](https://github.com/caikit/caikit-nlp/blob/main/Dockerfile) that builds an image which can be launched to start the `caikit.runtime` server(s) with the `caikit_nlp` modules configured.

Reviewer: ```suggestion
The `caikit-nlp` project provides a [Dockerfile](https://github.com/caikit/caikit-nlp/blob/main/Dockerfile) that builds an image that can be launched to start the `caikit.runtime` server(s) with the `caikit_nlp` modules configured.
```
